Adaptation to the Peng et Al framework

matol-16 opened the pull request.

Ayoub-Gourari committed:
Evaluate and save maximal gradient variance

Ayoub-Gourari committed:
Name adaptation

Ayoub-Gourari committed:
Plots of different scatterings, variances and feature mean

Ayoub-Gourari committed:
Fixed issues raised by Linh Vu

Ayoub-Gourari committed:
Fixed introduced models to yield probabilities rather than logits

Ayoub-Gourari committed:
Implementation of the static and dynamic label flippings

Ayoub-Gourari committed:
poisonned client in the training loop

Ayoub-Gourari committed:
Official LFighter implementation acting on weights

matol-16 commented on byzfl/attacks/attacks.py:
Il me semble que le static label flipping le correspond pas exactement à celui du papier *mean is more robust*. Dans mean is more robust, le flip est déterministe:  targets = 9 - targets. ON pourrait peut être ajouter une variable "random" pour effectuer un flip aléatoire dans un deuxième temps.

Ayoub-Gourari replied:
Fait !

matol-16 commented on byzfl/benchmark/evaluate_results.py:
Ici il me semble que dans chacune des méthodes que tu implémentes, tu plottes encore le accuracy path  (via tab_acc)? Par ailleurs, par exemple quand tu plots  la feature variance, pas besoin de le faire pour chaque attaque ou aggrégateur puisque a priori la feature variance dépend uniquement de la distribution. ON peut en discuter sur whatsapp. Concernant le rho (pas encore implémenté de ce que je comprends), je pense qu'on peut reprendre directement les formules du papier.

Ayoub-Gourari replied:
Oui c'est ça. Par rapport au plot de la feature variance, on peut appeler la fonction une fois par distribution (il me semble que nous qui choisiront quand on va les appeler dans l'autre repo). Concernant les convergence rates, j'attends qu'on avance et qu'on se mette d'accords sur la partie théorique pour pouvoir les implémenter, normalement les papiers introduisants chaque aggregateur robuste évaluent son rho.

matol-16 replied:
Mais donc je ne comprenais exactement pourquoi tu replottais à chaque fois le accuracy path, même dans la fonction de plot de la la feature mean: plot_maximum_regular_feature_mean

matol-16 commented on byzfl/benchmark/train.py:
```diff
@@ -166,22 +166,29 @@ def start_training(params):
+            "milestones": params_manager.get_milestones(),
+            "learning_rate_decay": params_manager.get_learning_rate_decay(),
+            "LabelFlipping": "LabelFlipping" == params_manager.get_attack_name(),
+            "training_dataloader": client_dataloaders[i],
```
ici, il me semble que le poisoned cleints récupère les mêmes dataloaders que les honest clients. Il faudrait probablement créer nb_honest+nb_byz_clients dataloaders, et assigner les nb_byz_clients derniers dataloaders aux clients empoisonnés

matol-16 replied:
Ca me fait penser: est-ce que on a pris en compte dans le reste de byzfl qu'on considère 2 types de clients désormais, et pas uniquement les honest clients soumis à des attaques omniscientes ? J'imagine qu'il y a des spillovers notamment pour les plots: il faudra pouvoir différencier les points des byzantins et des honnêtes au moment de plotter.

Ayoub-Gourari replied:
Il me semble que ce n'est pas grave, la seule différence c'est dans le calcul des "scatterings" qui est gérée dans la ligne 413 -> 419 du fichier train. (On a pas besoin de prendre en compte la norme des moyennes des features, ni de leur accuracies dans nos expériences)

matol-16 replied:
Ca marche ! en revanche il me semble qu'on a quand même le soucis des dataloaders partagés non ? (cf mon 1er commentaire). Genre le datadistributor va générer nb_honest dataloaders. Il faudrait probablement changer cela à nb_honest+nb_poisonned pour avoir bien une distribution différente pour chaque client.

Ayoub-Gourari replied:
Oui pardon je l'ai résolu. La class de data loader prends en arguments mtn le nombre total de workers

matol-16 commented on byzfl/benchmark/train.py:
```diff
@@ -209,19 +217,19 @@ def start_training(params):
 
-    if training_algorithm_name not in ["DSGD", "FedAvg"]:
+    if training_algorithm_name not in ["DSGD"]:
         raise ValueError(f"Training algorithm {training_algorithm_name} not supported, supported algorithms are 'DSGD' and 'FedAvg'")
```
message d'erreur à changer probabelemnt

Ayoub-Gourari replied:
Pas forcément je crois. Y'a moyen qu'on revienne à "FedAvg" pour LFighter (cf. commentaire sur LFighter sur le markdown des tâches)

matol-16 commented on byzfl/benchmark/train.py:
```diff
@@ -263,81 +271,71 @@ def start_training(params):
-            )
+            # Apply poisonning attack
+            for poisonned_client in poisonned_clients:
+                poisonned_client.compute_gradients()
```
ici, ca veut dire qu'on n'enregistre pas les variances et la loss des byzantins ?

Ayoub-Gourari replied:
On a pas besoin de la loss des byzantins (même pas de celle des honnêtes en vrai). Pour les variances je viens de régler.

matol-16 commented on byzfl/fed_framework/models.py:
ici, il faudrait peut être changer les noms des modèles ajoutés (type: CNN_MNIST_PENG) ou un truc comme ca, pcq il y a des modèles implémentés par byzfl avec le même nom mais en miiuscule (genre cnn_mnist)

matol-16 commented on byzfl/benchmark/train.py:
est-ce que tu penses qu'il serait utile d'ajouter un boolean pour gérer l'enregistrement des variances etc ? Je me dis que quand on passera sur des modèles plus gros, le faire à chaque training step sera coûteux

Ayoub-Gourari committed:
Fixed the staticlabel flipping, Included poisonned workers in gradient variance

Ayoub-Gourari committed:
changed models names

Ayoub-Gourari committed:
Fix poisonned data distributors

slivering commented on byzfl/attacks/attacks.py:
```diff
@@ -1121,11 +1122,33 @@ def __call__(self, honest_vectors):
+    
+    def __call__(self, model, inputs, outputs):      
+        classes = list(set(outputs))
+        mapping = {outputs : model.nb_classes - 1 - outputs for outputs in classes}
```
Ok, mais on pourrait aussi rendre la permutation configurable dans `__init__` (spécifier `mapping` en argument de l'attaque)

slivering commented on byzfl/attacks/attacks.py:
```diff
@@ -1121,11 +1122,33 @@ def __call__(self, honest_vectors):
+        classes = list(set(outputs))
+        mapping = {outputs : model.nb_classes - 1 - outputs for outputs in classes}
+
+        flipped_outputs = [mapping[output] for output in outputs[:int(self.p*len(outputs))]] + outputs[int(self.p*len(outputs)):]
```
Ok, mais on pourrait aussi effectuer le label flipping sur un sous-ensemble **aléatoire** des indices avec `np.random.Generator.choice(len(outputs), int(p * len(outputs)))`

(`np.random.Generator` : à configurer avec une seed)

slivering commented on byzfl/benchmark/managers.py:
```diff
@@ -111,18 +111,18 @@ def save_honest_scattering(self, honest_scattering_list, training_seed, data_dis
         np.savetxt(file_path, honest_scattering_list, fmt="%.4f", delimiter=",")
 
-    def save_byzantine_scattering(self, byzantine_scattering_list, training_seed, data_dist_seed):
+    def save_poisonned_scattering(self, poisonned_scattering_list, training_seed, data_dist_seed):
```
Un seul n à poisoned (à appliquer à chaque instance du mot)

slivering commented on byzfl/attacks/attacks.py:
```diff
@@ -1121,11 +1122,33 @@ def __call__(self, honest_vectors):
+    def __init__(self, p):
+        self.p = p
+
+    def __call__(self, model, inputs, outputs):
```
Cette méthode n'utilise pas le paramètre `self.p`...

slivering commented on byzfl/fed_framework/poisoning_client.py:
```diff
@@ -189,20 +194,17 @@ def apply_attack(self, inputs, targets):
-
-        n = len(inputs)
-        m = int(self.p * n)
-        poisoned_indices = np.random.choice(n, m, replace=False)
```
Pourquoi avoir déplacé ce code dans chaque attaque par label flipping ?
L'intérêt de ces lignes était d'appliquer l'attaque par empoisonnement sur tous les `poisoned_indices`. Cela évitait de dupliquer cette logique.

slivering commented on byzfl/aggregators/aggregators.py:
```diff
@@ -675,46 +676,101 @@ def clusters_dissimilarity(self, clusters):
-    def __call__(self, vectors):
-        m = len(vectors)
-        dw = [[] for _ in range(m)]
+    # Get average weights
```
`@staticmethod` pour appeler `LFighter.average_weights(w, marks)`.

slivering commented on byzfl/aggregators/aggregators.py:
```diff
@@ -675,46 +676,101 @@ def clusters_dissimilarity(self, clusters):
+        return w_avg
+
+    # Adapt to gradients ...
+    def aggregate(self, global_model, local_models, ptypes):
```
Non applicable car l'on ne dispose pas de state dicts mais de gradients 1D. Il faudra implémenter `__call__` en adaptant l'implémentation originale.

matol-16 committed:
minor syntax error fix - mathias

matol-16 committed:
fix attack parameters for poisoning

matol-16 committed:
fix softmax model : add dimension to softmax function

matol-16 committed:
fix gradient computation : take the mean

matol-16 committed:
fix: string of honest gradient plots caused bugs

matol-16 committed:
fix gradient computation result formatting (else, array vs sequence error)

matol-16 committed:
fix: gradients was a list, not a tensor

matol-16 committed:
fix: minor syntax error on gradient_variance

matol-16 committed:
fix: faba was bugged: call syntax + remain was not a tensor

matol-16 committed:
fix: incompatibility main vs variance_and_plot in the calculation of the scattering list

Ayoub-Gourari committed:
Added a function to do 1 training step for Lfighter

Ayoub-Gourari committed:
Lfighter aggregator using weights directly

matol-16 committed:
fix: delete poisoning attacks in the gradient attack file #7

matol-16 committed:
Merge branch 'variance_and_plot' of https://github.com/EA-Robust-AI-X2023/byzfl into variance_and_plot

matol-16 committed:
fix: model names in small caps #5

Ayoub-Gourari committed:
Add poisonned workers heterogeneities

Ayoub-Gourari committed:
added variables to controle the measures to do inside a client

Ayoub-Gourari committed:
Fixed models names

Ayoub-Gourari committed:
template of the functions reading the controle variables for the measures

Ayoub-Gourari committed:
Added measures controle variables in training loop

matol-16 committed:
fix: plot of gradient scatterings. Regrouped honest and poisoned gradients scatterings, adapted plots to the peng et al style

matol-16 committed:
fix: naming error in save mean feature norm

matol-16 committed:
Merge branch 'variance_and_plot' of https://github.com/EA-Robust-AI-X2023/byzfl into variance_and_plot

matol-16 committed:
added gradient variance/scattering + feature measures as options (+minor fix of these options in train.py

matol-16 commented on byzfl/aggregators/aggregators.py:
```diff
@@ -691,7 +691,7 @@ def average_weights(w, marks):
 
     # Adapt to gradients ...
-    def aggregate(self, global_model, local_models, ptypes):
+    def __call__(self, global_model, local_models):
```
Ici, tu calls avec global_model + local_models en paramètre. Ca ne fonctionne pas avec train.py, qui appelle uniquement un aggregateur avec en argument un objet "vector" (les params du modèles, ou les gradients).

Il faut donc probabelment faire une exception pour Lfighter, pour pouvoir lui feed les modèles locaux et globaux... On peut peut être s'inspirer de ce que font Peng et al

Ayoub-Gourari replied:
J'ai push une modification dans une nouvelle branche Lfighter_adaptation (désolé, on travaillais au même temps et du coup je suis devenu asynchro).

matol-16 committed:
fix dynamic label flipping #16 #12

matol-16 committed:
minor syntax fixes #14

Ayoub-Gourari committed:
Changed Lfighter's arguments to models instead of weights or clients

matol-16 committed:
add the possibility to scatter momentums or raw gradients #17

matol-16 committed:
Merge remote-tracking branch 'origin/main' into variance_and_plot

matol-16 committed:
fix syntax errors

matol-16 committed:
fix: syntax error in the data distributor

matol-16 committed:
fix Lfighter #13

matol-16 committed:
Merge branch 'Lfighter_adaptation' into variance_and_plot #13

matol-16 committed:
fix #13 merge issues

matol-16 committed:
fix : measure gradient variance and scattering periodically, rather than every train step #14

matol-16 committed:
fix: ensure that gradients are calculated and store for the hole batch, not just the last point in the batch

matol-16 committed:
fix big models #15

matol-16 committed:
fix: adapt static label flipping to the case where we use a gpu #14

matol-16 committed:
fix: momentum scattering as an option

matol-16 committed:
add min_size as a param for modified distributions

matol-16 committed:
fix: accuracy curves too tight

matol-16 committed:
fix: gpu robust calculation of scatterings

matol-16 committed:
fix: data distribution bugged for low alphas

matol-16 committed:
modify scattering plots: plot all attacks together (cf peng et al)

matol-16 committed:
fix: error in condition handling in client

matol-16 committed:
fix: error in attack file (no ident for Noattack)

matol-16 committed:
Major fix: modified distributions were yielding wrong indexes

slivering commented on byzfl/fed_framework/client.py:
```diff
@@ -123,14 +154,55 @@ def _backward_pass(self, inputs, targets, train_acc=False):
+                loss = self.criterion(output, target.unsqueeze(0))  # Add batch dimension
+                loss.backward()
+                
+                grad_vector = torch.cat([p.grad.view(-1) for p in self.model.parameters() if p.grad is not None])
```
`grad_vector = self.get_flat_gradients()` devrait être équivalent

slivering reviewed (commented):
Quelques remarques principales :

1. Laisser la responsabilité à `PoisoningClient` de sélectionner une fraction **aléatoire** des données à empoisonner (rendre la seed configurable)
2. Calcul douteux de `dw` et `db` dans LFighter
3. `Server.update_model_with_models` suspicieux (problème de signature avec les agrégateurs qui prennent des gradients en argument)

Ayoub-Gourari committed:
Fixed Lfighter's issue

matol-16 committed:
add plotting of worker distribution feature

matol-16 committed:
Merge branch 'variance_and_plot' of https://github.com/EA-Robust-AI-X2023/byzfl into variance_and_plot

matol-16 committed:
fix: dirichelt modified takes the train set size rather than the total size

matol-16 committed:
fix plot of per worker class distribution

slivering committed:
Refactor LFighter

slivering committed:
Refactor FABA

slivering committed:
Use loguru instead of print

slivering committed:
Refactor StaticLabelFlipping

matol-16 committed:
add the measure of exclusivty + saving the partition

slivering committed:
Update dependencies

slivering committed:
Fix update_model_with_models for LFighter

slivering committed:
Use dict gradients to factor in momentum

slivering committed:
FIx typos "poisonned" -> "poisoned"

matol-16 committed:
fix faba device bug

matol-16 committed:
fix: import error

matol-16 committed:
fix faba device bug

matol-16 committed:
plot of exclusivity measure

slivering committed:
fix: reconciliate LFighter and momentum

Two steps to be consistent with the paper:

1. apply LFighter on the gradients w/o momentum to identify poisoned gradients
2. aggregate the honest gradients w/ momentum

matol-16 committed:
fix : lfighter last issues after test

matol-16 committed:
Merge branch 'variance_and_plot' into refactoring

matol-16 committed:
Merge pull request #21 from EA-Robust-AI-X2023/refactoring after test

Refactoring

matol-16 committed:
Merge branch 'variance_and_plot' of https://github.com/EA-Robust-AI-X2023/byzfl into variance_and_plot

matol-16 committed:
fix: syntax error

matol-16 committed:
add randomness in label separation

matol-16 committed:
fix exclusivity measure

matol-16 committed:
fix: no softmax in the forward pass of softmax regression

matol-16 committed:
fix training pipeline: forgot to update byzantine client!

matol-16 committed:
refactor training pipeline

matol-16 committed:
add lfighter option to chose momentum gradients

matol-16 committed:
fix: record also byzantine client measures

matol-16 committed:
add min accuracy option to accuracy plots

matol-16 committed:
add subset size plot

matol-16 committed:
add  boxplots

matol-16 committed:
add entropy measure + p value to the boxplots
